src/plugins/kibana_utils/docs/state_sync/on_fly_state_migrations.md

```diff
@@ -0,0 +1,46 @@
+# On-fly state migrations
+
+When retrieving initial state from storage we shouldn't forget about possible outdated state.
+Similar to [handling initial state](./initial_state.md) example, applications could handle migrations during initialisation.
+
+```ts
+import { migrate } from '../app/state_helpers';
+const urlStateStorage = createKbnUrlStateStorage();
+const initialStateFromUrl = urlStateStorage.get('_a');
+
+// merge default state and initial state and migrate it to the current version
+const initialState = migrate({
+  ...defaultState,
+  ...initialStateFromUrl,
+});
+
+const stateContainer = createStateContainer(initialState);
+```
+
+It is also possible to apply migrations for any incoming state similar example in [handling empty or incomplete state](./empty_or_incomplete_incoming_state.md).
+
+```ts
+import { migrate } from '../app/state_helpers';
+
+const urlStateStorage = createKbnUrlStateStorage();
+const initialStateFromUrl = urlStateStorage.get('_a');
+
+// merge together default state and initial state and migrate it to current version if needed
```

Comment:
merge ~~together~~ default state and initial state and migrate ~~it~~ them to current version, if needed